File: ios/LittleDill/ArenaExpression.swift
import SwiftUI

enum ArenaSmile: String, CaseIterable {
    case smile, beam, smirk, dimples, grin, crooked

    init(variety:String) {
        self = ["dill":.smile,"gherkin":.beam,"garlic":.smirk,"butter":.dimples,"chili":.grin,"pepper":.crooked][variety] ?? .smile
    }

    func draw(_ context:GraphicsContext,r:Double,lineWidth:Double) {
        var c = context
        c.scaleBy(x:r,y:r)
        let ink = GraphicsContext.Shading.color(DillTheme.ink)
        let stroke = StrokeStyle(lineWidth:lineWidth/r,lineCap:.round,lineJoin:.round)
        var path = Path()
        switch self {
        case .smile:
            path.move(to:CGPoint(x:-0.13,y:0.19))
            path.addQuadCurve(to:CGPoint(x:0.13,y:0.19),control:CGPoint(x:0,y:0.42))
            c.stroke(path,with:ink,style:stroke)
        case .beam:
            path.move(to:CGPoint(x:-0.15,y:0.19))
            path.addQuadCurve(to:CGPoint(x:0.15,y:0.19),control:CGPoint(x:0,y:0.24))
            path.addQuadCurve(to:CGPoint(x:0,y:0.4),control:CGPoint(x:0.14,y:0.43))
            path.addQuadCurve(to:CGPoint(x:-0.15,y:0.19),control:CGPoint(x:-0.14,y:0.43))
            c.fill(path,with:ink)
        case .smirk:
            path.move(to:CGPoint(x:-0.13,y:0.26))
            path.addQuadCurve(to:CGPoint(x:0.17,y:0.16),control:CGPoint(x:0.03,y:0.35))
            path.move(to:CGPoint(x:0.135,y:0.15)); path.addLine(to:CGPoint(x:0.19,y:0.17))
            c.stroke(path,with:ink,style:stroke)
        case .dimples:
            path.move(to:CGPoint(x:-0.18,y:0.23))
            path.addQuadCurve(to:CGPoint(x:0,y:0.23),control:CGPoint(x:-0.09,y:0.38))
            path.addQuadCurve(to:CGPoint(x:0.18,y:0.23),control:CGPoint(x:0.09,y:0.38))
            c.stroke(path,with:ink,style:stroke)
        case .grin:
            path.move(to:CGPoint(x:-0.19,y:0.2)); path.addLine(to:CGPoint(x:0.19,y:0.2))
            path.addQuadCurve(to:CGPoint(x:0,y:0.4),control:CGPoint(x:0.13,y:0.43))
            path.addQuadCurve(to:CGPoint(x:-0.19,y:0.2),control:CGPoint(x:-0.13,y:0.43))
            c.fill(path,with:ink)
            c.fill(Path(CGRect(x:-0.13,y:0.205,width:0.26,height:0.055)),with:.color(Color(hex:0xF8F6ED)))
        case .crooked:
            path.move(to:CGPoint(x:-0.16,y:0.16))
            path.addQuadCurve(to:CGPoint(x:0.13,y:0.28),control:CGPoint(x:-0.05,y:0.39))
            path.move(to:CGPoint(x:-0.195,y:0.175)); path.addLine(to:CGPoint(x:-0.135,y:0.15))
            c.stroke(path,with:ink,style:stroke)
        }
    }
}

struct ArenaPickupCadence {
    struct Cue {
        let at:Double
        let index:Int
        let streak:Int
        var volume:Double {0.7 * (1 - Double(streak) * 0.08)}
        var sound:DillSound {[.nibble,.nibble2,.nibble3,.nibble4][index]}
    }
    private(set) var previous:Cue?

    mutating func next(at now:Double) -> Cue? {
        let streak = previous.map {now - $0.at <= 1 ? min(5,$0.streak + 1) : 0} ?? 0
        if let previous, now - previous.at < 0.3 + Double(min(4,previous.streak)) * 0.04 {return nil}
        let cue = Cue(at:now,index:previous.map {($0.index + 1) % 4} ?? 0,streak:streak)
        previous = cue
        return cue
    }
}

File: arena-web/arena-expression.mjs
export const EAT_OVERLAP = .6;
export const SMILES = Object.freeze({ dill: 'smile', gherkin: 'beam', garlic: 'smirk', butter: 'dimples', chili: 'grin', pepper: 'crooked' });
export const skinSmile = variety => SMILES[variety] || SMILES.dill;

export function pickupCue(previous, now) {
  const streak = previous && now - previous.at <= 1 ? Math.min(5, previous.streak + 1) : 0;
  if (previous && now - previous.at < .3 + Math.min(4, previous.streak) * .04) return null;
  const index = previous ? (previous.index + 1) % 4 : 0;
  return { at: now, index, streak, volume: .7 * (1 - streak * .08) };
}

export function drawSmile(ctx, variety, size) {
  ctx.save(); ctx.scale(size, size); ctx.lineWidth /= size; ctx.beginPath();
  switch (skinSmile(variety)) {
    case 'beam':
      ctx.moveTo(-.15, .19); ctx.quadraticCurveTo(0, .24, .15, .19);
      ctx.quadraticCurveTo(.14, .43, 0, .4); ctx.quadraticCurveTo(-.14, .43, -.15, .19);
      ctx.fill(); break;
    case 'smirk':
      ctx.moveTo(-.13, .26); ctx.quadraticCurveTo(.03, .35, .17, .16); ctx.stroke();
      ctx.beginPath(); ctx.moveTo(.135, .15); ctx.lineTo(.19, .17); ctx.stroke(); break;
    case 'dimples':
      ctx.moveTo(-.18, .23); ctx.quadraticCurveTo(-.09, .38, 0, .23);
      ctx.quadraticCurveTo(.09, .38, .18, .23); ctx.stroke(); break;
    case 'grin':
      ctx.moveTo(-.19, .2); ctx.lineTo(.19, .2); ctx.quadraticCurveTo(.13, .43, 0, .4);
      ctx.quadraticCurveTo(-.13, .43, -.19, .2); ctx.fill();
      ctx.save(); ctx.fillStyle = '#F8F6ED'; ctx.fillRect(-.13, .205, .26, .055); ctx.restore(); break;
    case 'crooked':
      ctx.moveTo(-.16, .16); ctx.quadraticCurveTo(-.05, .39, .13, .28); ctx.stroke();
      ctx.beginPath(); ctx.moveTo(-.195, .175); ctx.lineTo(-.135, .15); ctx.stroke(); break;
    default:
      ctx.moveTo(-.13, .19); ctx.quadraticCurveTo(0, .42, .13, .19); ctx.stroke();
  }
  ctx.restore();
}
